hbase-server/src/main/java/org/apache/hadoop/hbase/rsgroup/RSGroupAdminEndpoint.java

```diff
@@ -108,85 +108,119 @@ RSGroupAdminServiceImpl getGroupAdminService() {
       if (groupNameOfNs == null || groupNameOfNs.equals(RSGroupInfo.DEFAULT_GROUP)) {
         // do not check for default group
-        return true;
+        return;
       }
       rsGroupInfo = groupAdminServer.getRSGroupInfo(groupNameOfNs);
       if (rsGroupInfo == null) {
         throw new ConstraintException("RSGroup " + groupNameOfNs + " for table " +
-            desc.getTableName() + "(inherit from namespace) does not exist");
+          desc.getTableName() + "(inherit from namespace) does not exist");
       }
     }
-    return master.getServerManager().createDestinationServersList().stream()
-        .anyMatch(onlineServer -> rsGroupInfo.containsServer(onlineServer.getAddress()));
+    if (rsGroupInfo.getServers().isEmpty()) {
+      throw new ConstraintException(
+        "No servers in the rsgroup " + rsGroupInfo.getName() + " for " + desc);
+    }
   }
 
   @Override
   public void preCreateTableAction(ObserverContext<MasterCoprocessorEnvironment> ctx,
-      TableDescriptor desc, RegionInfo[] regions) throws IOException {
+    TableDescriptor desc, RegionInfo[] regions) throws IOException {
     checkGroupExists(desc.getRegionServerGroup());
-    if (!desc.getTableName().isSystemTable() && !rsgroupHasServersOnline(desc)) {
-      throw new HBaseIOException("No online servers in the rsgroup for " + desc);
-    }
+    checkForEmptyGroup(desc);
   }
 
   @Override
   public TableDescriptor preModifyTable(ObserverContext<MasterCoprocessorEnvironment> ctx,
-      TableName tableName, TableDescriptor currentDescriptor, TableDescriptor newDescriptor)
-      throws IOException {
-    checkGroupExists(newDescriptor.getRegionServerGroup());
+    TableName tableName, TableDescriptor currentDescriptor, TableDescriptor newDescriptor)
+    throws IOException {
+    if (!currentDescriptor.getRegionServerGroup().equals(newDescriptor.getRegionServerGroup())) {
+      RSGroupInfo group = checkGroupExists(newDescriptor.getRegionServerGroup());
+      if (group != null && group.getServers().isEmpty()) {
+        throw new ConstraintException(
+          "No servers in the rsgroup " + group.getName() + " for " + newDescriptor);
+      }
+    }
     return MasterObserver.super.preModifyTable(ctx, tableName, currentDescriptor, newDescriptor);
   }
 
+  private void checkNamespaceGroup(NamespaceDescriptor ns) throws IOException {
+    String groupName = ns.getConfigurationValue(RSGroupInfo.NAMESPACE_DESC_PROP_GROUP);
+    if (groupName == null) {
+      return;
+    }
+    RSGroupInfo group = groupAdminServer.getRSGroupInfo(groupName);
+    if (group == null) {
+      throw new ConstraintException(
+        "RSGroup " + groupName + " for namespace " + ns.getName() + " does not exist");
+    }
+    if (group.getServers().isEmpty()) {
+      throw new ConstraintException(
+        "No servers in the rsgroup " + group.getName() + " for namespace " + ns.getName());
+    }
+  }
+
```

Comment:
I thought these method can be refacotred to 3 methods:
1. getRSGroupForTable()
2. getRSGroupForNamespace()
3. checkRSGroup() : it will check RSGroup exists and not empty.

Reply:
The logic here is not that straight-forward. If we are modifying table, we do not need to check the group if the group config is not changed or we just removed the group config. So does namespace...